lib/vdsm/storage/hsm.py

```diff
@@ -2871,6 +2871,9 @@ def prune_bitmaps(self, sdUUID, imgUUID, volUUID, baseUUID):
             baseUUID (str): The UUID of the volume that will get its bitmaps
                 pruned if missing or invalid in top.
         """
+        vars.task.getSharedLock(STORAGE, imgUUID)
+        vars.task.getSharedLock(STORAGE, volUUID)
+        vars.task.getExclusiveLock(STORAGE, baseUUID)
```

Comment:
This not how storage locks are taken, see how we lock the image and the volume in the sp.py.

Reply:
So more specific advice - we cannot take volume locks, since locking the image take a lock on all volumes:

I think it works like this:
1. sd.py _registerResourceNamespaces register imageResourcesNamespace with imageResourceFactory
2. imageResourceFactory takes a lock on all volume of the image when creating a resource
3. So when you take a lock on an image, you lock all the volume in this image with the same lock type

Since we want an exclusive lock on the base volume, we must take exclusive lock on the image, which will take also exclusive lock on the top volume.

So we need something like:

    img_ns = rm.getNamespace(sc.IMAGE_NAMESPACE, sdUUID)
    with rm.acquireResource(img_ns, imgUUID, rm.EXCLUSIVE):
        purne bitmaps here...

BUT: this must be AFTER we take the storage domain lock, otherwise we can have a deadlock. All locks must be taken in the same order:

1. sc.STORAGE
2. sc.IMAGE_NAMESPACE
3. sc.VOLUME_NAMESPACE
4. sc.LVM_ACTIVATION_NAMESPACE

This is why the constants are named: "00_storage", "01_img", "02_vol", "03_lvm".

Code taking multiple locks must sort the locks to ensure the order.